File: day-06/day-06.py
#!/usr/bin/env python
################################################################################
# Copyright (C) 2019 Michael Wiseman                                           #
#                                                                              #
# This program is free software: you can redistribute it and/or modify it      #
# under the terms of the GNU Affero General Public License as published by the #
# Free Software Foundation, either version 3 of the License, or (at your       #
# option) any later version.                                                   #
#                                                                              #
# This program is distributed in the hope that it will be useful, but WITHOUT  #
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or        #
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License #
# for more details.                                                            #
#                                                                              #
# You should have received a copy of the GNU Affero General Public License     #
# along with this program.  If not, see <https://www.gnu.org/licenses/>.       #
################################################################################
import networkx as nx
import pprint

pp = pprint.PrettyPrinter(indent=4)

# Function definition

def part_one(puzz_input):
    orbits = nx.DiGraph()

    for l in puzz_input:
        p, s = l.split(')')
        orbits.add_edge(p, s)
    print(nx.transitive_closure(orbits).size())

def part_two(puzz_input):
    orbits = nx.Graph()

    for l in puzz_input:
        p, s = l.split(')')
        orbits.add_edge(p, s)
    print(nx.shortest_path_length(orbits, 'YOU', 'SAN') - 2)

# Main program

fllst = open('day-06-input.txt', 'r').read().splitlines()

part_one(fllst) # 261306
part_two(fllst) # 382

File: day-06/day-06.pl
#!/usr/bin/env perl
################################################################################
# Copyright (C) 2019 Michael Wiseman                                           #
#                                                                              #
# This program is free software: you can redistribute it and/or modify it      #
# under the terms of the GNU Affero General Public License as published by the #
# Free Software Foundation, either version 3 of the License, or (at your       #
# option) any later version.                                                   #
#                                                                              #
# This program is distributed in the hope that it will be useful, but WITHOUT  #
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or        #
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License #
# for more details.                                                            #
#                                                                              #
# You should have received a copy of the GNU Affero General Public License     #
# along with this program.  If not, see <https://www.gnu.org/licenses/>.       #
################################################################################
use Array::Utils qw(:all);
use common::sense;
use Data::Dumper;
use Graph;
use Graph::Directed;
use Graph::TransitiveClosure;
use v5.30;

# Function definition

sub part_one
{
    my @puzz_input = @{$_[0]};
    my $orbits = Graph::Directed->new;

    for my $line (@puzz_input)
    {
        my ($p, $s) = split /\)/, $line;
        $orbits->add_edge($p, $s);
    }
    say scalar $orbits->TransitiveClosure_Floyd_Warshall;
}

sub part_two
{
    my @puzz_input = @{$_[0]};
    my $orbits = Graph->new;

    for my $line (@puzz_input)
    {
        my ($p, $s) = split /\)/, $line;
        $orbits->add_edge($p, $s);
    }
    my @s = $orbits->SP_Dijkstra('COM', 'SAN');
    my @y = $orbits->SP_Dijkstra('COM', 'YOU');
    say((scalar array_diff(@s, @y)) - 2);
}

# Open input file

my $filename = 'day-06-input.txt';
open(my $fh, '<', $filename);
my @fl = <$fh>;
close $fh;
chomp(@fl);

# Main program

part_one(\@fl); # 261306
part_two(\@fl); # 382
